### PR TITLE
Search bug fixes

### PR DESCRIPTION
**Pull request by indhumeyyappan**
small bugs like  output columns in where clause does not return all records, distinct clause in a select query, hack form response display size is small, JSON array display does not have commas,

### CODE REVIEW

File: Paco-Server/ear/default/web/js/controllers.js [hunk 1]
```diff
@@ -722,63 +722,58 @@ pacoApp.controller('HackCtrl', ['$scope', '$http', function ($scope, $http) {
...
   $scope.errorResponse = '';
   $scope.submitForm = function () {
-    $http({
-      method: 'POST',
-      url: '/csSearch',
-      data: angular.fromJson($scope.hackRequest),
-      headers: {'Content-Type': 'application/json'}
+     $http({
+     method  : 'POST',
+     url     : '/csSearch?tz=America/Los_Angeles',
```
[the remark this answers is not in the document]
Comment (BobEvans) [reply Paco-Server/ear/default/web/js/controllers.js]: Also, if we want to use the client side timezone, we either need to use offsets instead of ID for timezone, or, we might want to investigate the moments.js library which seems to offer IDs back. Not sure if that is worth exploring. Either, way it should probably be a separate feature branch.

File: Paco-Server/ear/default/web/partials/hack.html [hunk 2]
```diff
@@ -27,8 +28,8 @@
         onLoad: aceResLoaded,
         onChange: aceResChanged,
         rendererOptions: {
-          minLines: 10,
-          maxLines: 10
+          minLines: 30,
+          maxLines: 30
```
Comment (BobEvans) [inline Paco-Server/ear/default/web/partials/hack.html]: More viewing area - Yay!

File: Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java [hunk 3]
```diff
@@ -122,6 +122,7 @@ public void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOEx
       resp.setContentType("text/plain");
```
Comment (BobEvans) [inline Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java]: Is the output actually text/plain? Isn't it application/json?

File: Paco-Server/ear/default/web/js/controllers.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (BobEvans) [reply Paco-Server/ear/default/web/js/controllers.js]: Actually, skip the moments.js we should be able to use the javascript api: Intl.DateTimeFormat().resolvedOptions().timeZone

File: Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java (hunk 3 above)
Comment (BobEvans) [reply Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java]: Looks like this comment was missed. I looked it up as well. In this case, the ContentType should indeed be "application/json".

File: Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java [hunk 4]
```diff
@@ -158,11 +159,14 @@ public void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOEx
           SearchUtil.addJoinClause(clientJsqlStatement);
         }
         if(sqlQueryObj.isFullEventAndOutputs()) {
-          optimizedSelect = modifyToOptimizePerformance(clientJsqlStatement);
+          boolean outputColsInWhere = false;
```
Comment (BobEvans) [inline Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java]: Small suggestion: these three statements could be one statement.
Original
---
 boolean outputColsInWhere = false;
 +          if (sqlQueryObj.getCriteriaQuery().contains(OutputBaseColumns.ANSWER) || sqlQueryObj.getCriteriaQuery().contains(OutputBaseColumns.NAME)) {
 +            outputColsInWhere = true;
 +          }

suggestion =>

 boolean outputColsInWhere = sqlQueryObj.getCriteriaQuery().contains(OutputBaseColumns.ANSWER) ||    
      sqlQueryObj.getCriteriaQuery().contains(OutputBaseColumns.NAME));

File: Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java [hunk 5]
```diff
@@ -272,12 +278,16 @@ private static Select modifyToOptimizePerformance(Select currentSelect) throws J
...
      modifyProjectionColumnsToClientQuery(currentSelect);
      clientReqQuery.setSelectBody(currentSelect.getSelectBody());
+     
      clientReqQuery.setAlias(new Alias(CLIENT_REQUEST));
      
      // second join on condition
-     secondJoinCondition.append(EventServerColumns.TABLE_NAME + "." + Constants.UNDERSCORE_ID+ " = " +CLIENT_REQUEST + "." + Constants.UNDERSCORE_ID + " AND ");
-     secondJoinCondition.append(OutputBaseColumns.TABLE_NAME + "." + OutputBaseColumns.EVENT_ID + " = "  + CLIENT_REQUEST + "." + Constants.UNDERSCORE_ID + " AND ");
-     secondJoinCondition.append(OutputBaseColumns.TABLE_NAME + "." + OutputBaseColumns.NAME+ " = "  + CLIENT_REQUEST + "." + OutputBaseColumns.NAME);
+     secondJoinCondition.append(EventServerColumns.TABLE_NAME + "." + Constants.UNDERSCORE_ID+ " = " +CLIENT_REQUEST + "." + Constants.UNDERSCORE_ID + AND);
```
Comment (BobEvans) [inline Paco-Server/src/com/google/sampling/experiential/server/CloudSqlSearchServlet.java]: nit: please add spaces between elements in concatenation where missing

File: Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java [hunk 6]
```diff
@@ -108,10 +108,18 @@ public Builder fullEventAndOutputs(Boolean fullEventAndOutputs) {
...
+      // find if there is a distinct clause
+      boolean isDistinct = false;
+      for(String s : obj.getProjection()) {
+        if (s.startsWith(Constants.DISTINCT)) {
+          isDistinct = true;
+          break;
+        }
+      }
+      // provide default sort order which is Event._Id desc
+      if (obj.getGroupBy() == null && obj.sortOrder == null && !isDistinct) {
```
Comment (BobEvans) [inline Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java]: micro-optimization: if it is the case that obj.getGroupBy != null || obj.sortOrder != null then we do not need to check the projections for a distinct. Is that correct? I imagine this is not expensive so I wouldn't change it just for this case, I just wanted to make sure that I understand it correctly.

File: Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java [hunk 7]
```diff
@@ -108,10 +108,18 @@ public Builder fullEventAndOutputs(Boolean fullEventAndOutputs) {
...
+        obj.sortOrder = EventBaseColumns.TABLE_NAME+"."+Constants.UNDERSCORE_ID.concat(Constants.BLANK).concat(Constants.DESC);
+      }
 
       if (obj.getProjection() == null) {
```
Comment (BobEvans) [inline Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java]: Will obj.getProjection() return null, or, will it return an empty list? If it can return null, the iteration on line 113 to check for a DISTINCT needs to test for null.

Comment (indhumeyyappan) [reply Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java]: yep...you are right. I missed to check that.

File: Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java (hunk 6 above)
Comment (indhumeyyappan) [reply Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java]: I am glad you asked the question. It was a great catch.

File: Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java [hunk 8]
```diff
@@ -108,6 +108,11 @@ public Builder fullEventAndOutputs(Boolean fullEventAndOutputs) {
...
 
     private Builder addDefaultValues(SQLQuery obj) {
+      // default projection is *
+      if (obj.getProjection() == null) {
+        obj.projection = new String[] { Constants.STAR };
+        obj.fullEventAndOutputs = true;
+      } 
       // find if there is a distinct clause
       boolean isDistinct = false;
       for(String s : obj.getProjection()) {
```
Comment (BobEvans) [inline Shared/src/com/pacoapp/paco/shared/model2/SQLQuery.java]: The for block will throw a NullPointerException i the obj.getProtection() returns null.
